src/coreclr/vm/assembly.cpp

```diff
@@ -598,29 +603,55 @@ Assembly *Assembly::CreateDynamic(AppDomain *pDomain, CreateDynamicAssemblyArgs
     }
 
     NewHolder<DomainAssembly> pDomainAssembly;
+    BOOL                      usesAllocatorFromBinderContext = FALSE;
 
     {
         GCX_PREEMP();
 
+        AssemblyLoaderAllocator* pAssemblyLoaderAllocator = nullptr;
+
+        if (pBinderContext != nullptr)
+        {
+            pBinderContext->GetLoaderAllocator((LPVOID*)&pAssemblyLoaderAllocator);
+            usesAllocatorFromBinderContext = TRUE;
+        }
+
         // Create a new LoaderAllocator if appropriate
         if ((args->access & ASSEMBLY_ACCESS_COLLECT) != 0)
         {
-            AssemblyLoaderAllocator *pAssemblyLoaderAllocator = new AssemblyLoaderAllocator();
-            pAssemblyLoaderAllocator->SetCollectible();
-            pLoaderAllocator = pAssemblyLoaderAllocator;
-
-            // Some of the initialization functions are not virtual. Call through the derived class
-            // to prevent calling the base class version.
-            pAssemblyLoaderAllocator->Init(pDomain);
+            if (pAssemblyLoaderAllocator == nullptr || !pAssemblyLoaderAllocator->IsCollectible())
+            {
+                pAssemblyLoaderAllocator = new AssemblyLoaderAllocator();
+                pAssemblyLoaderAllocator->SetCollectible();
+
+                // Some of the initialization functions are not virtual. Call through the derived class
+                // to prevent calling the base class version.
+                pAssemblyLoaderAllocator->Init(pDomain);
+
+                // Setup the managed proxy now, but do not actually transfer ownership to it.
+                // Once everything is setup and nothing can fail anymore, the ownership will be
+                // atomically transfered by call to LoaderAllocator::ActivateManagedTracking().
+                pAssemblyLoaderAllocator->SetupManagedTracking(&args->loaderAllocator);
+                usesAllocatorFromBinderContext = FALSE;
+            }
 
-            // Setup the managed proxy now, but do not actually transfer ownership to it.
-            // Once everything is setup and nothing can fail anymore, the ownership will be
-            // atomically transfered by call to LoaderAllocator::ActivateManagedTracking().
-            pAssemblyLoaderAllocator->SetupManagedTracking(&args->loaderAllocator);
+            pLoaderAllocator = pAssemblyLoaderAllocator;
         }
         else
         {
-            pLoaderAllocator = pDomain->GetLoaderAllocator();
+            if (pAssemblyLoaderAllocator == nullptr)
+            {
+                pLoaderAllocator = pDomain->GetLoaderAllocator();
+                usesAllocatorFromBinderContext = TRUE;
```

Comment:
This is weird - the names are confusing for this case. Maybe it would be better to rename `usesAllocatorFromBinderContext` to something like `usesExistingAllocator` (which would then imply to suppress release), or the very specific `releaseAllocator` (and negate the value).

Reply:
Actually - I would swap the meaning and call it `createdNewAllocator` (or something similar) - and below where we check for `args->access & ASSEMBLY_ACCESS_COLLECT` I would only check for this new `createdNewAllocator`. The checks for the `ASSEMBLY_ACCESS_COLLECT` are not semantically right (it's not that we want the assembly to be collectibale, we need to do additional things if we created a new allocator).

Reply:
Fixed